### DISCUSSION TITLE
TextPlugin - Set Default Font for Whole Project

### DISCUSSION THREAD

**Discussion by onkoe** [bevyengine/bevy#11839] [Q&A] [answered]
Hey there! In #8445, a technique to change the font [is mentioned](https://github.com/bevyengine/bevy/pull/8445#issuecomment-1515817076). 

I don't particularly love this method, mostly because I want to alter the TextPlugin's defaults instead of checking to see if there's a font I should replace each frame. 

Are there any methods to change the default font that are more like the following for 0.12.1? 

```rust
use bevy::{prelude::*, text::TextPlugin};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins.set(TextPlugin.with_some_config(ThatChangesTheFont::new())))
        .run();
}
```

[`TextPlugin`](https://docs.rs/bevy/latest/bevy/text/struct.TextPlugin.html) is just a unit struct, so I'm not exactly sure how to feed it a [`TextStyle`](https://docs.rs/bevy/latest/bevy/text/struct.TextStyle.html#).

## Official Support

The [Bevy 0.11 release notes](https://bevyengine.org/news/bevy-0-11/#default-font) mention that you can now change the default font, but doesn't exactly state how to do so. 

Does anyone have experience with setting the global default font?

**Comment by rlidwka** [comment]
> instead of checking to see if there's a font I should replace each frame

If I understood the example correctly, you don't need to do that *each frame*, it's only one time when your font is loaded.

**Reply by onkoe** [reply]
You might be right! I'm not sure what the `Update` schedule means on the low-level side of things. 😄️

If it does run each frame, then I would assume that at least the `.run_if()` part does a small check on each schedule loop. That's nothing horrible, though it does feel 'off' to me. By any chance, do you know why the technique is set up that way?

**Reply by rparrett** [reply, marked as answer]
Bevy itself is embedding a tiny stripped-down font in the binary, bypassing the asset server. But the example you linked is making the assumption that you want to load a normal font the normal way through the asset server and not bloat up your binary. In that case, you need to poll until the font is actually loaded (an asynchronous process that may take more than one frame) before you can replace it.

If you're okay with embedding the font, I think you can simplify like this:

```rust
use bevy::{asset::load_internal_binary_asset, prelude::*};

fn main() {
    let mut app = App::new();

    app.add_plugins(DefaultPlugins).add_systems(Startup, setup);

    // This needs to happen after `DefaultPlugins` is added.
    load_internal_binary_asset!(
        app,
        TextStyle::default().font,
        "../assets/fonts/FiraSans-Bold.ttf",
        |bytes: &[u8], _path: String| { Font::try_from_bytes(bytes.to_vec()).unwrap() }
    );

    app.run();
}
```

**Reply by rlidwka** [reply]
> By any chance, do you know why the technique is set up that way?

Executing the system once when font loads requires a lot more code (see libraries like `iyes_progress`).

So if your goal is to make a small demo for github, it is a reasonable thing to go with quick and dirty solution of executing it each frame. But in real application, there would be separate `Loading` state that handles all of it.

**Reply by onkoe** [reply]
Thanks for sharing your experience! That code snippet was just what I was looking for.

I also appreciate the idea of a `Loading` state. I hadn't yet considered that side of things, but that implementation does make sense for a "demo" showing folks that the font can be altered.

**Reply by doceazedo** [reply]
note: `TextStyle` has been renamed to `TextFont` in bevy >= 0.15
